### ISSUE TITLE
could not find application file: ranch.app using docker

### ISSUE THREAD

**Issue by teodor-pripoae** [phoenixframework/phoenix#637]
Hi,

I'm trying to setup a phoenix project using docker. Everything works fine if starting from a clean pull, but if the application gets compiled in a container and then start another container using the same project directory, I'm getting an error:

```
** (Mix) Could not start application ranch: could not find application file: ranch.app
```

Steps for reproducing the error:

Start one container, mounting myapp to /code in the container.
Commands starting with $ are run on the host, commands starting with > are run inside the container
```

$ docker pull trenpixster/elixir:1.0.2
$ docker run -i -t -v ~/code/myapp/:/code -e ERLANG_COOKIE=abcdefgh trenpixster/elixir:1.0.2 /bin/bash

> cd /code
> mix local.hex --force
> rm -r deps _build
> mix deps.get
> mix deps.compile
> iex --name "foo@container1" --cookie $ERLANG_COOKIE -S mix phoenix.server
```

Then, start another container, sharing myapp inside /code

```
$ docker run -i -t -v ~/code/myapp/:/code -e ERLANG_COOKIE=abcdefgh trenpixster/elixir:1.0.2 /bin/bash

>  iex --name "foo@container2" --cookie $ERLANG_COOKIE -S mix phoenix.server

** (Mix) Could not start application ranch: could not find application file: ranch.app
```

If I remove deps and _build inside the second container and build again, everything works as expected inside the second container. Next, if I stop the first container iex session and start again (files are synced, we now have the build from second container inside container 1) everything works again. So, is there any state outside project directory which should be present ?

**tl;dr** I guess this can be also reproduced like this: compile app on computer A. Copy all project directory including deps and _build to computer B. Attempt to start server, then you get this error

**Comment by amagdas** [comment]
I think you should use releases, I'm using Exrm: https://hex.pm/packages/exrm

**Comment by amagdas** [comment]
Follow this tutorial as well: https://github.com/lancehalvorsen/phoenix-guides/blob/master/K_deployment.md

**Comment by ericmj** [comment]
> tl;dr I guess this can be also reproduced like this: compile app on computer A. Copy all project directory including deps and _build to computer B. Attempt to start server, then you get this error

Can you actually reproduce this? I can't.

**Comment by teodor-pripoae** [comment]
@amagdas I'm trying to do it for development, not for packaging for production.

I think I fixed the issue by running `mix deps.get` again.

```
root@cad852bd0c43:/code# mix deps.get
Dependency resolution completed successfully
All dependencies up to date
```

I'm not sure why it does need this command to be run if we already have the deps in the `deps` directory. Also, I was having the system clock for the vagrant vm before system clock on my computer, and getting this errors. Not sure if running `mix deps.get` or setting up my system clock solved this issue.

```
warning: mtime (modified time) for "/code/_build/development/lib/ws_service/ebin/sha1.beam" was set to the future, resetting to now
warning: mtime (modified time) for "/code/_build/development/lib/ws_service/.compile.elixir" was set to the future, resetting to now
warning: mtime (modified time) for "/code/_build/development/lib/ws_service/.compile.erlang" was set to the future, resetting to now
warning: mtime (modified time) for "/code/_build/development/lib/ws_service/ebin/ws_service.app" was set to the future, resetting to now
```

Also, it seems to compile ranch and a few others even if the code is already compiled in `_build` directory.

```
root@cad852bd0c43:/code# iex --name "foo@macbookpro" -S mix phoenix.server
Erlang/OTP 17 [erts-6.2] [source] [64-bit] [async-threads:10] [kernel-poll:false]

==> ranch (compile)
==> rabbit_common (compile)
==> amqp_client (compile)
==> cowlib (compile)
... rest of my app files
```

**Comment by josevalim** [comment]
The system clock is definitely the issue. Rebar or Elixir may have failed to compile part of the code because the future timestamp told there was no need to compile some particular bits. In other cases, the timestamp will always be behind, causing it to always recompile.

**Comment by josevalim** [comment]
@teodor-pripoae maybe you can touch the files after you copy them to the VM? It will at least guarantee they have a reasonable timestamp.

**Comment by teodor-pripoae** [comment]
I did a ntpclient update, but this didn't solve the issue. The other issue was that I was missing the `~/.hex` directory containing hex registry stuff, since the container was immutable, each time I was starting a new container I needed to run deps.get.

Please consider this issue solved since it was a fault in my configuration.
